solidity/contracts/bridge/Frauds.sol

```diff
@@ -0,0 +1,577 @@
+                    signature.s
+                )
+            )
+        );
+
+        FraudChallenge storage challenge = self.challenges[challengeKey];
+
+        require(challenge.reportedAt > 0, "Fraud challenge does not exist");
+        require(!challenge.closed, "Fraud challenge closed");
+
+        // Ensure SIGHASH_ALL type was used during signing, which is represented
+        // by type value `1`.
+        require(extractSighashType(preimage) == 1, "Wrong sighash type");
+
+        return
+            witness
+                ? extractUtxoKeyFromWitnessPreimage(preimage)
+                : extractUtxoKeyFromNonWitnessPreimage(preimage);
+    }
+
+    /// @notice Finalizes fraud challenge defeat by marking a pending fraud
+    ///         challenge against the wallet as closed and sending the ether
+    ///         deposited by the challenger to the treasury.
+    ///         In order to finalize the challenge defeat the same
+    ///         `walletPublicKey` and signature (represented by `r`, `s` and
+    ///         `v`) must be provided as were used in the fraud challenge.
+    ///         Additionally a preimage must be provided which was used to
+    ///         calculate the sighash during input signing.
+    /// @param walletPublicKey The public key of the wallet in the uncompressed
+    ///        and unprefixed format (64 bytes)
+    /// @param preimage The preimage which produces sighash used to generate the
+    ///        ECDSA signature that is the subject of the fraud claim. It is a
+    ///        serialized subset of the transaction. The exact subset used as
+    ///        the preimage depends on the transaction input the signature is
+    ///        produced for. See BIP-143 for reference
+    /// @param signature Bitcoin signature in the R/S/V format
+    /// @param treasury Treasury associated with the Bridge
+    /// @dev Should be called for a fraud challenge defeat attempt that has been
+    ///      verified.
+    function defeatChallenge(
+        Data storage self,
+        bytes memory walletPublicKey,
+        bytes memory preimage,
+        Frauds.RSVSignature calldata signature,
+        address treasury
+    ) external {
+        bytes32 sighash = preimage.hash256();
+
+        uint256 challengeKey = uint256(
+            keccak256(
+                abi.encodePacked(
+                    walletPublicKey,
+                    sighash,
+                    signature.v,
+                    signature.r,
+                    signature.s
+                )
+            )
+        );
+
+        FraudChallenge storage challenge = self.challenges[challengeKey];
+
+        // Close fraud challenge as it was successfully defeated
+        challenge.closed = true;
+
+        // Send the ether deposited by the challenger to the treasury
+        /* solhint-disable avoid-low-level-calls */
+        // slither-disable-next-line low-level-calls
+        treasury.call{gas: 5000, value: challenge.depositAmount}("");
```

Comment:
Can we set this limit much higher? 5000 today should be more than enough but if Ethereum forks and increases cost of opcodes, we do not want to redeploy everything. Let's do it even 100k. 

Same for `notifyFraudChallengeDefeatTimeout`.

Reply:
Increased to `100k` in 33628a5de32bdc2d80bd3a9ce451ba3cd5f39873.